### PR TITLE
ref #278: enable configurable variable(s) for the less compiler

### PR DESCRIPTION
**Pull request by UlrichKu**
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed issues  | chameleon-system/chameleon-system#278
| License       | MIT

The change to an upgrade or changelog file is missing as probably "UPGRADE-6.3.md" might not be the correct one.
(Change/added feature and changed constructor.)

### CODE REVIEW

**Comment by pixeljunkie** [comment]
@UlrichKu Could you please fix the merge conflicts here.

**Comment by UlrichKu** [comment]
Also added upgrade notice.

File: UPGRADE-7.0.md [hunk 1]
```diff
@@ -19,6 +19,12 @@ project requires WYSIWYG JavaScript, allow it by setting the configuration key
 
 # Changed Features
 
+## Less Compilation supports STATIC_CONTENT_URL
```
Comment (umulmrum) [inline UPGRADE-7.0.md]:
```suggestion
## Less Compilation Supports STATIC_CONTENT_URL
```

File: UPGRADE-7.0.md [hunk 2]
```diff
@@ -19,6 +19,12 @@ project requires WYSIWYG JavaScript, allow it by setting the configuration key
...
+
+The less compiler TPkgViewRendererLessCompiler now supports an additional argument for variables that are passed to
+less files. Currently this is used to inject STATIC_CONTENT_URL which can be used to load resources from another system
+(ie a CDN). This in turn can be configured with `chameleon_system_view_renderer.less_compiler: static_content_url`.
+
```
Comment (umulmrum) [inline UPGRADE-7.0.md]: Could you move the new sections to a new file `UPGRADE-7.1md`?

File: UPGRADE-7.0.md [hunk 3]
```diff
@@ -19,6 +19,12 @@ project requires WYSIWYG JavaScript, allow it by setting the configuration key
 
 # Changed Features
 
+## Less Compilation supports STATIC_CONTENT_URL
+
+The less compiler TPkgViewRendererLessCompiler now supports an additional argument for variables that are passed to
+less files. Currently this is used to inject STATIC_CONTENT_URL which can be used to load resources from another system
+(ie a CDN). This in turn can be configured with `chameleon_system_view_renderer.less_compiler: static_content_url`.
```
Comment (umulmrum) [inline UPGRADE-7.0.md]:
```suggestion
(e.g. a CDN). This in turn can be configured with `chameleon_system_view_renderer.less_compiler: static_content_url`.
```

File: src/ViewRendererBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -22,6 +22,9 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->defaultValue('chameleon/outbox/static/less')
                     ->info('Path used as working and output directory for the less compiler, relative to the web root directory. CAUTION: All files in this directory are deleted on cache clear.')
                 ->end()
+                ->scalarNode('static_content_url')
+                    ->defaultValue('')
+                    ->info('URL for static content (images, fonts, ...) if served from a different location (i. e. a CDN). Can be omitted. Default is empty.')
```
Comment (umulmrum) [inline src/ViewRendererBundle/DependencyInjection/Configuration.php]:
```suggestion
                    ->info('URL for static content (images, fonts, ...) if served from a different location (e.g. a CDN). Can be omitted. Default is empty.')
```

File: src/ViewRendererBundle/DependencyInjection/ChameleonSystemViewRendererExtension.php [hunk 5]
```diff
@@ -28,5 +28,8 @@ protected function loadInternal(array $mergedConfig, ContainerBuilder $container
 
         $lessCompiler = $container->getDefinition('chameleon_system_view_renderer.less_compiler');
         $lessCompiler->replaceArgument(0, $mergedConfig['css_dir']);
+        $lessCompiler->replaceArgument(2, ['STATIC_CONTENT_URL' => '"'.$mergedConfig['static_content_url'].'"']);
+        // NOTE must be included in .less something like: url("@{STATIC_CONTENT_URL}")
```
Comment (umulmrum) [inline src/ViewRendererBundle/DependencyInjection/ChameleonSystemViewRendererExtension.php]: Both comment lines are quite unclear to me. Could you try and rephrase?

File: src/ViewRendererBundle/objects/helper/TPkgViewRendererLessCompiler.php [hunk 6]
```diff
@@ -23,15 +23,22 @@ class TPkgViewRendererLessCompiler
...
      */
     private $resourceCollectionRefreshPrefix;
 
-    public function __construct(string $cssDirRelativeToWebRoot, string $resourceCollectionRefreshPrefix)
+    /**
+     * @var array
+     */
+    private $additionalVariables;
+
+    public function __construct(string $cssDirRelativeToWebRoot, string $resourceCollectionRefreshPrefix, array $additionalVariables)
```
Comment (umulmrum) [inline src/ViewRendererBundle/objects/helper/TPkgViewRendererLessCompiler.php]: We could think about adding a method `addLessVariable(string $key, string $value)` instead. Although constructor injection in general is preferred, we would gain a few advantages:
- In projects or third-party bundles additional variables can be defined through compiler passes.
- Users of the class don't need to learn how the array needs to look like; they can pass key and value as separate method arguments. Also type-safety is improved a bit.
- No BC break, so we could think about adding this feature to 6.2.x (only if "necessary" - as this is a feature, I'd prefer master).

Comment (UlrichKu) [reply src/ViewRendererBundle/objects/helper/TPkgViewRendererLessCompiler.php]: Now using a method addAdditionalVariables().

File: UPGRADE-7.1.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
+UPGRADE FROM 7.0 TO 7.1
```
Comment (pixeljunkie) [inline UPGRADE-7.1.md]: This file needs to be removed and the text needs to be added to the UPGRADE-7.0.md

Comment (UlrichKu) [reply UPGRADE-7.1.md]: ok

**Pull request merged**